### ISSUE TITLE
Can't explicit load many-to-many relationships.

### ISSUE THREAD

**Issue by CheloXL** [dotnet/efcore#24329]
## Ask a question

I'm trying to explicit-load a many to many relationship and it is not working as I expect. I'm using the following configuration.
Seed data is correctly added through EF.

```C#
public class ManyM_DB
{
	public string Id { get; set; } = Guid.NewGuid().ToString("N");
	public ICollection<ManyN_DB> ManyN_DB { get; set; }
	public ICollection<ManyMN_DB> ManyNM_DB { get; set; }
}

public class ManyN_DB
{
	public string Id { get; set; } = Guid.NewGuid().ToString("N");
	public ICollection<ManyM_DB> ManyM_DB { get; set; }
	public ICollection<ManyMN_DB> ManyNM_DB { get; set; }
}

public sealed class ManyMN_DB
{
	public string Id { get; set; } = Guid.NewGuid().ToString("N");

	public string ManyM_Id { get; set; }
	public ManyM_DB ManyM_DB { get; set; }

	public string ManyN_Id { get; set; }
	public ManyN_DB ManyN_DB { get; set; }
}

internal sealed class ManyMnConfiguration : IEntityTypeConfiguration<ManyM_DB>
{
	public void Configure(EntityTypeBuilder<ManyM_DB> builder)
	{
		builder.ToTable("many_m");
		builder.HasKey(b => b.Id);

		builder
			.HasMany(x => x.ManyN_DB)
			.WithMany(x => x.ManyM_DB)
			.UsingEntity<ManyMN_DB>(
				j => j
					.HasOne(x => x.ManyN_DB)
					.WithMany(x => x.ManyNM_DB)
					.HasForeignKey(x => x.ManyN_Id),
				j => j
					.HasOne(x => x.ManyM_DB)
					.WithMany(x => x.ManyNM_DB)
					.HasForeignKey(x => x.ManyM_Id)
			);
		}
}

internal sealed class ManyNConfiguration : IEntityTypeConfiguration<ManyN_DB>
{
	public void Configure(EntityTypeBuilder<ManyN_DB> builder)
	{
		builder.ToTable("many_n");
		builder.HasKey(b => b.Id);
	}
}

internal sealed class ManyMNConfiguration : IEntityTypeConfiguration<ManyMN_DB>
{
	public void Configure(EntityTypeBuilder<ManyMN_DB> builder)
	{
		builder.ToTable("many_mn");
		builder.HasKey(b => b.Id);
	}
}
```

What I'm doing is loading ManyM_DB and later trying to load it's related N entities, with:

```C#
var m = Context.Find<ManyM_DB>(dbState.Id);
await Context.Entry(m).Collection(x => x.ManyN_DB).LoadAsync();
```

At this point I expect the property `ManyN_DB` to be filled in with related data, but I've got null.
If I first do `await Context.Entry(m).Collection(x => x.ManyNM_DB).LoadAsync();` _then_ the above `LoadAsync` works.

Using MySql and obtaining the query through `var query = Context.Entry(app).Collection(x => x.ManyNM_DB).Query().ToQueryString();` returns the following query (ids are examples).

```SQL
SELECT `t`.`Id`, `m`.`Id`, `t`.`Id0`, `t0`.`Id`, `t0`.`Id0`
FROM `many_m` AS `m`
INNER JOIN (
    SELECT `m1`.`Id`, `m0`.`Id` AS `Id0`, `m0`.`ManyM_Id`
    FROM `many_mn` AS `m0`
    LEFT JOIN `many_n` AS `m1` ON `m0`.`ManyN_Id` = `m1`.`Id`
) AS `t` ON `m`.`Id` = `t`.`ManyM_Id`
LEFT JOIN (
    SELECT `m3`.`Id`, `m2`.`Id` AS `Id0`, `m2`.`ManyN_Id`
    FROM `many_mn` AS `m2`
    LEFT JOIN `many_m` AS `m3` ON `m2`.`ManyM_Id` = `m3`.`Id`
    WHERE `m3`.`Id` = '552521170365997056'
) AS `t0` ON `t`.`Id` = `t0`.`ManyN_Id`
WHERE `m`.`Id` = '552521170365997056'
ORDER BY `m`.`Id`, `t`.`Id0`, `t`.`Id`, `t0`.`Id0`, `t0`.`Id`
```
Running this query returns the right data. But of course the above query only return indices, not records.

Relevant query log from my tests:
```
Debug: XUnitLogger: Context 'TestDbContext' started tracking 'ManyM_DB' entity with key '{Id: 552521170365997056}'.
Debug: XUnitLogger: Compiling query expression: 
'DbSet<ManyM_DB>()
    .AsTracking()
    .Where(e => EF.Property<string>(e, "Id") == __p_0)
    .SelectMany(e => e.ManyN_DB)
    .NotQuiteInclude(e => e.ManyM_DB
        .Where(e => EF.Property<string>(e, "Id") == __p_0))'
Debug: XUnitLogger: Including navigation: 'ManyN_DB.ManyM_DB'.
Debug: XUnitLogger: Generated query execution expression: 
'queryContext => new QueryingEnumerable<ManyN_DB>(
    queryContext, 
    InMemoryShapedQueryCompilingExpressionVisitor.Table(
        queryContext: queryContext, 
        entityType: EntityType: ManyM_DB)
        .Where(valueBuffer => ExpressionExtensions.ValueBufferTryReadValue<string>(
            valueBuffer: valueBuffer, 
            index: 0, 
            property: Property: ManyM_DB.Id (string) Required PK AfterSave:Throw) == InMemoryExpressionTranslatingExpressionVisitor.GetParameterValue<string>(
            queryContext: queryContext, 
            parameterName: "__p_0"))
        .SelectMany(
            collectionSelector: valueBuffer => InMemoryShapedQueryCompilingExpressionVisitor.Table(
                queryContext: queryContext, 
                entityType: EntityType: ManyMN_DB)
                .Where(valueBuffer => IsFalse((ValueBufferTryReadValue(valueBuffer, 0, Property: ManyM_DB.Id (string) Required PK AfterSave:Throw) == null))
                 && object.Equals(
                    objA: ExpressionExtensions.ValueBufferTryReadValue<string>(
                        valueBuffer: valueBuffer, 
                        index: 0, 
                        property: Property: ManyM_DB.Id (string) Required PK AfterSave:Throw), 
                    objB: ExpressionExtensions.ValueBufferTryReadValue<string>(
                        valueBuffer: valueBuffer, 
                        index: 1, 
                        property: Property: ManyMN_DB.ManyM_Id (string) FK Index)))
                .GroupJoin(
                    inner: InMemoryShapedQueryCompilingExpressionVisitor.Table(
                        queryContext: queryContext, 
                        entityType: EntityType: ManyN_DB), 
                    outerKeySelector: valueBuffer => ExpressionExtensions.ValueBufferTryReadValue<string>(
                        valueBuffer: valueBuffer, 
                        index: 2, 
                        property: Property: ManyMN_DB.ManyN_Id (string) FK Index), 
                    innerKeySelector: valueBuffer => ExpressionExtensions.ValueBufferTryReadValue<string>(
                        valueBuffer: valueBuffer, 
                        index: 0, 
                        property: Property: ManyN_DB.Id (string) Required PK AfterSave:Throw), 
                    resultSelector: (outer, inner) => new TransparentIdentifier<ValueBuffer, IEnumerable<ValueBuffer>>(
                        Outer = outer, 
                        Inner = inner
                    ))
                .SelectMany(
                    collectionSelector: collection => collection.Inner
                        .DefaultIfEmpty(new ValueBuffer(new object[]{ null })), 
                    resultSelector: (outer, inner) => new ValueBuffer(new object[]
                    { 
                        ExpressionExtensions.ValueBufferTryReadValue<string>(
                            valueBuffer: outer.Outer, 
                            index: 0, 
                            property: Property: ManyMN_DB.Id (string) Required PK AfterSave:Throw), 
                        ExpressionExtensions.ValueBufferTryReadValue<string>(
                            valueBuffer: outer.Outer, 
                            index: 1, 
                            property: Property: ManyMN_DB.ManyM_Id (string) FK Index), 
                        ExpressionExtensions.ValueBufferTryReadValue<string>(
                            valueBuffer: outer.Outer, 
                            index: 2, 
                            property: Property: ManyMN_DB.ManyN_Id (string) FK Index), 
                        ExpressionExtensions.ValueBufferTryReadValue<string>(
                            valueBuffer: inner, 
                            index: 0, 
                            property: Property: ManyN_DB.Id (string) Required PK AfterSave:Throw) 
                    })), 
            resultSelector: (outer, inner) => new ValueBuffer(new object[]
            { 
                ExpressionExtensions.ValueBufferTryReadValue<string>(
                    valueBuffer: outer, 
                    index: 0, 
                    property: Property: ManyM_DB.Id (string) Required PK AfterSave:Throw), 
                ExpressionExtensions.ValueBufferTryReadValue<string>(
                    valueBuffer: inner, 
                    index: 3, 
                    property: Property: ManyN_DB.Id (string) Required PK AfterSave:Throw) 
            }))
        .Select(valueBuffer => new ValueBuffer(new object[]
        { 
            ExpressionExtensions.ValueBufferTryReadValue<string>(
                valueBuffer: valueBuffer, 
                index: 1, 
                property: Property: ManyN_DB.Id (string) Required PK AfterSave:Throw), 
            InMemoryShapedQueryCompilingExpressionVisitor.Table(
                queryContext: queryContext, 
                entityType: EntityType: ManyMN_DB)
                .Where(valueBuffer => IsFalse((ValueBufferTryReadValue(valueBuffer, 1, Property: ManyN_DB.Id (string) Required PK AfterSave:Throw) == null))
                 && object.Equals(
                    objA: ExpressionExtensions.ValueBufferTryReadValue<string>(
                        valueBuffer: valueBuffer, 
                        index: 1, 
                        property: Property: ManyN_DB.Id (string) Required PK AfterSave:Throw), 
                    objB: ExpressionExtensions.ValueBufferTryReadValue<string>(
                        valueBuffer: valueBuffer, 
                        index: 2, 
                        property: Property: ManyMN_DB.ManyN_Id (string) FK Index)))
                .GroupJoin(
                    inner: InMemoryShapedQueryCompilingExpressionVisitor.Table(
                        queryContext: queryContext, 
                        entityType: EntityType: ManyM_DB), 
                    outerKeySelector: valueBuffer => ExpressionExtensions.ValueBufferTryReadValue<string>(
                        valueBuffer: valueBuffer, 
                        index: 1, 
                        property: Property: ManyMN_DB.ManyM_Id (string) FK Index), 
                    innerKeySelector: valueBuffer => ExpressionExtensions.ValueBufferTryReadValue<string>(
                        valueBuffer: valueBuffer, 
                        index: 0, 
                        property: Property: ManyM_DB.Id (string) Required PK AfterSave:Throw), 
                    resultSelector: (outer, inner) => new TransparentIdentifier<ValueBuffer, IEnumerable<ValueBuffer>>(
                        Outer = outer, 
                        Inner = inner
                    ))
                .SelectMany(
                    collectionSelector: collection => collection.Inner
                        .DefaultIfEmpty(new ValueBuffer(new object[]{ null })), 
                    resultSelector: (outer, inner) => new ValueBuffer(new object[]
                    { 
                        ExpressionExtensions.ValueBufferTryReadValue<string>(
                            valueBuffer: outer.Outer, 
                            index: 0, 
                            property: Property: ManyMN_DB.Id (string) Required PK AfterSave:Throw), 
                        ExpressionExtensions.ValueBufferTryReadValue<string>(
                            valueBuffer: outer.Outer, 
                            index: 1, 
                            property: Property: ManyMN_DB.ManyM_Id (string) FK Index), 
                        ExpressionExtensions.ValueBufferTryReadValue<string>(
                            valueBuffer: outer.Outer, 
                            index: 2, 
                            property: Property: ManyMN_DB.ManyN_Id (string) FK Index), 
                        ExpressionExtensions.ValueBufferTryReadValue<string>(
                            valueBuffer: inner, 
                            index: 0, 
                            property: Property: ManyM_DB.Id (string) Required PK AfterSave:Throw) 
                    }))
                .Where(valueBuffer => ExpressionExtensions.ValueBufferTryReadValue<string>(
                    valueBuffer: valueBuffer, 
                    index: 3, 
                    property: Property: ManyM_DB.Id (string) Required PK AfterSave:Throw) == InMemoryExpressionTranslatingExpressionVisitor.GetParameterValue<string>(
                    queryContext: queryContext, 
                    parameterName: "__p_0"))
                .Select(valueBuffer => new ValueBuffer(new object[]{ ExpressionExtensions.ValueBufferTryReadValue<string>(
                    valueBuffer: valueBuffer, 
                    index: 3, 
                    property: Property: ManyM_DB.Id (string) Required PK AfterSave:Throw) })) 
        })), 
    Func<QueryContext, ValueBuffer, ManyN_DB>, 
    EfRepoTests.Context.TestDbContext, 
    False
)'
```

EF Core version: 5.0.3
Database provider: InMemory (but also tried MySQL)
Target framework: (e.g. .NET 5.0)
Operating system: Windows 10
IDE: Visual Studio 2019 16.9

**Comment by AndriySvyryd** [comment]
@CheloXL You can make this work by removing `ManyMN_DB.Id` and relying on the PK created by convention (`ManyM_Id, ManyN_Id`)

**Comment by CheloXL** [comment]
@AndriySvyryd will try that, but I need that ID because I'm using that intermediate table (with additional data, not shown here).

**Comment by ajcvickers** [comment]
@CheloXL As a workaround, mark the FK properties in the join table as required. For example:

```C#
builder
    .HasMany(x => x.ManyN_DB)
    .WithMany(x => x.ManyM_DB)
    .UsingEntity<ManyMN_DB>(
        j => j
            .HasOne(x => x.ManyN_DB)
            .WithMany(x => x.ManyNM_DB)
            .HasForeignKey(x => x.ManyN_Id),
            .IsRequired(),
        j => j
            .HasOne(x => x.ManyM_DB)
            .WithMany(x => x.ManyNM_DB)
            .HasForeignKey(x => x.ManyM_Id)
            .IsRequired());
```

@smitpatel @maumar I have investigated this further and I think it is a query bug, or at least a deficiency in the loading query. It happens when the FK properties in the join table are optional. So this works:

```C#
public sealed class ManyMN_DB
{
    public int Id { get; set; }

    public int ManyM_Id { get; set; }
    public ManyM_DB ManyM_DB { get; set; }

    public int ManyN_Id { get; set; }
    public ManyN_DB ManyN_DB { get; set; }
}
```

But this does not:

```C#
public sealed class ManyMN_DB
{
    public int Id { get; set; }

    public int? ManyM_Id { get; set; }
    public ManyM_DB ManyM_DB { get; set; }

    public int? ManyN_Id { get; set; }
    public ManyN_DB ManyN_DB { get; set; }
}
```

When loading we generate a query like this:

```C#
var loaded = context
    .Set<ManyM_DB>()
    .AsTracking()
    .Where(e => e.Id == m.Id)
    .SelectMany(e => e.ManyN_DB)
    .Include(e => e.ManyM_DB.Where(e => e.Id == m.Id))
    .ToList();
```

When the join table FKs are non-nullable (the working case) this generates the following SQL:

```sql
info: Database.Command[20101]
      Executed DbCommand (7ms) [Parameters=[@__p_0='1'], CommandType='Text', CommandTimeout='30']
      SELECT [t].[Id], [m].[Id], [t].[Id0], [t0].[Id], [t0].[ManyM_Id], [t0].[ManyN_Id], [t0].[Id0]
      FROM [ManyM_DB] AS [m]
      INNER JOIN (
          SELECT [m1].[Id], [m0].[Id] AS [Id0], [m0].[ManyM_Id]
          FROM [ManyMN_DB] AS [m0]
          INNER JOIN [ManyN_DB] AS [m1] ON [m0].[ManyN_Id] = [m1].[Id]
      ) AS [t] ON [m].[Id] = [t].[ManyM_Id]
      LEFT JOIN (
          SELECT [m2].[Id], [m2].[ManyM_Id], [m2].[ManyN_Id], [m3].[Id] AS [Id0]
          FROM [ManyMN_DB] AS [m2]
          INNER JOIN [ManyM_DB] AS [m3] ON [m2].[ManyM_Id] = [m3].[Id]
          WHERE [m3].[Id] = @__p_0
      ) AS [t0] ON [t].[Id] = [t0].[ManyN_Id]
      WHERE [m].[Id] = @__p_0
      ORDER BY [m].[Id], [t].[Id0], [t].[Id], [t0].[Id], [t0].[Id0]
```

But when the keys are nullable we generate this:

```sql
info: Database.Command[20101]
      Executed DbCommand (6ms) [Parameters=[@__p_0='1'], CommandType='Text', CommandTimeout='30']
      SELECT [t].[Id], [m].[Id], [t].[Id0], [t0].[Id], [t0].[Id0]
      FROM [ManyM_DB] AS [m]
      INNER JOIN (
          SELECT [m1].[Id], [m0].[Id] AS [Id0], [m0].[ManyM_Id]
          FROM [ManyMN_DB] AS [m0]
          LEFT JOIN [ManyN_DB] AS [m1] ON [m0].[ManyN_Id] = [m1].[Id]
      ) AS [t] ON [m].[Id] = [t].[ManyM_Id]
      LEFT JOIN (
          SELECT [m3].[Id], [m2].[Id] AS [Id0], [m2].[ManyN_Id]
          FROM [ManyMN_DB] AS [m2]
          LEFT JOIN [ManyM_DB] AS [m3] ON [m2].[ManyM_Id] = [m3].[Id]
          WHERE [m3].[Id] = @__p_0
      ) AS [t0] ON [t].[Id] = [t0].[ManyN_Id]
      WHERE [m].[Id] = @__p_0
      ORDER BY [m].[Id], [t].[Id0], [t].[Id], [t0].[Id0], [t0].[Id]
```

Notice that the FK values are missing from the top-level select. This results in the join entity not being materialized.

Test code that I have been using to investigate:

```C#
public class ManyM_DB
{
    public int Id { get; set; }
    public ICollection<ManyN_DB> ManyN_DB { get; set; }
    public ICollection<ManyMN_DB> ManyNM_DB { get; set; }
}

public class ManyN_DB
{
    public int Id { get; set; }
    public ICollection<ManyM_DB> ManyM_DB { get; set; }
    public ICollection<ManyMN_DB> ManyNM_DB { get; set; }
}

public sealed class ManyMN_DB
{
    public int Id { get; set; }

    public int ManyM_Id { get; set; }
    public ManyM_DB ManyM_DB { get; set; }

    public int ManyN_Id { get; set; }
    public ManyN_DB ManyN_DB { get; set; }
}

public class SomeDbContext : DbContext
{
    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        => optionsBuilder
            .UseSqlServer(Your.ConnectionString)
            .EnableSensitiveDataLogging();

    protected override void OnModelCreating(ModelBuilder modelBuilder)
    {
        modelBuilder.Entity<ManyM_DB>(
            builder =>
            {
                builder
                    .HasMany(x => x.ManyN_DB)
                    .WithMany(x => x.ManyM_DB)
                    .UsingEntity<ManyMN_DB>(
                        j => j
                            .HasOne(x => x.ManyN_DB)
                            .WithMany(x => x.ManyNM_DB)
                            .HasForeignKey(x => x.ManyN_Id),
                        j => j
                            .HasOne(x => x.ManyM_DB)
                            .WithMany(x => x.ManyNM_DB)
                            .HasForeignKey(x => x.ManyM_Id)
                    );
            });

        modelBuilder.Entity<ManyN_DB>(
            builder =>
            {
            });

        modelBuilder.Entity<ManyMN_DB>(
            builder =>
            {
            });
    }
}

[ConditionalFact]
public void Main()
{
    int id;
    using (var context = new SomeDbContext())
    {
        context.Database.EnsureDeleted();
        context.Database.EnsureCreated();

        var m = new ManyM_DB();
        var n = new ManyN_DB();
        context.AddRange(m, n);
        m.ManyN_DB = new List<ManyN_DB> { n };

        context.SaveChanges();

        id = m.Id;
    }

    // using (var context = new SomeDbContext())
    // {
    //     var m = context.Find<ManyM_DB>(1);
    //     var n = context.Find<ManyN_DB>(1);
    //     var nm = context.Find<ManyMN_DB>(1);
    //
    //     Assert.Equal(3, context.ChangeTracker.Entries().Count());
    //     Assert.Equal(1, m.ManyN_DB.Count);
    //     Assert.Equal(1, m.ManyN_DB.Single().ManyM_DB.Count);
    //     Assert.Equal(1, m.ManyNM_DB.Count);
    //     Assert.Equal(1, m.ManyN_DB.Single().ManyNM_DB.Count);
    // }

    using (var context = new SomeDbContext())
    {
        var m = context.Find<ManyM_DB>(id);

        var loaded = context
            .Set<ManyM_DB>()
            .AsTracking()
            .Where(e => e.Id == m.Id)
            .SelectMany(e => e.ManyN_DB)
            .Include(e => e.ManyM_DB.Where(e => e.Id == m.Id))
            .ToList();

        //context.Entry(m).Collection(x => x.ManyN_DB).Load();

        Assert.Equal(3, context.ChangeTracker.Entries().Count());
        Assert.Equal(1, m.ManyN_DB.Count);
        Assert.Equal(1, m.ManyN_DB.Single().ManyM_DB.Count);
        Assert.Equal(1, m.ManyNM_DB.Count);
        Assert.Equal(1, m.ManyN_DB.Single().ManyNM_DB.Count);
    }
}
```

**Comment by AndriySvyryd** [comment]
Should we patch this?

**Comment by smitpatel** [comment]
Can patch this. The fix is very low risk. Reopening so we can discuss in triage.

**Comment by ajcvickers** [comment]
Note from triage: given that this has only one customer report (no duplicates and no 👍 on the issue), we will not patch this yet. If you are hitting this and need it patched in 5.0, then please comment or file a new issue.
